### ISSUE TITLE
Expiring forced source fitting

### ISSUE THREAD

**Issue by gijzelaerr** [transientskp/tkp#456]
Currently when a source gets above the detection threshold the source is monitored until infinity, also when it doesn't appear anymore or was an artefact. This may get problematic in the long run. 

We need to come op with an expiration policy and (configurable) procedure.

**Comment by AntoniaR** [comment]
What about just keeping the policy very simple and stop after 'x' non-detections following the last detection of the source? Note this is not the total number of non-detections for a source as that would cause problems... Where x is a user configurable parameter and x=0 gives the current situation where sources are tracked forever? That could both solve this issue but keep TraP sufficiently flexible for astronomers. However, I am not sure if this would require major code or database changes in TraP.

These sources would still need to be included in the source association step after they have fallen off the list of sources for monitoring, as they might re-appear again and this is important to know about.

Also, each time the source is detected, the counter for x needs to restart - so the source is monitored for a short time after each time it is blindly detected.

**Comment by gijzelaerr** [comment]
had a sit down with bart, we think this should work:

* add a `forcedfits_count` column to `runningcatalog` table with default value 0: https://github.com/transientskp/tkp/blob/master/tkp/db/sql/statements/tables/runningcatalog.sql#L21

* add a argument `forcedfits_max` to `get_nulldetections`, add a check to the query that filter the result on if the `forcedfits_count` column value of runningcatalog table is below `forcedfits_max` .  https://github.com/transientskp/tkp/blob/master/tkp/db/nulldetections.py#L15

* in the `associate_nd` function in `nulldetections.py` we need to add a new query that will increment the forcedfits_count column of a runningcatalog entry for every forced fits. https://github.com/transientskp/tkp/blob/master/tkp/db/nulldetections.py#L78

* in `_update_1_to_1_runcat` in `associations.py` we should add a statement that always resets `forcedfits_count` to 0. https://github.com/transientskp/tkp/blob/master/tkp/db/associations.py#L1480

* we don't neeed to update the `_update_1_to_many_runcat` query since we have a default value column of 0: https://github.com/transientskp/tkp/blob/master/tkp/db/associations.py#L872

**Comment by timstaley** [comment]
I'll not have time to dig through the SQL myself for a couple of weeks, but it sounds like you guys have been through it with a fine-tooth comb, thanks. Does @bartscheers have time to implement this, or are you planning on getting stuck in yourself @gijzelaerr ?

**Comment by gijzelaerr** [comment]
Maybe. Not sure if I have the time also. But we at least have a better idea
of the effort this will require.

Op wo 9 sep. 2015 13:59 schreef Tim Staley <notifications@github.com>:

**Comment by bartscheers** [comment]
I can implement this, but that will be first week of October.

**Comment by gijzelaerr** [comment]
i'm already working on it, you can review the PR when it is finished if you
want.

2015-09-25 13:19 GMT+02:00 Bart Scheers <notifications@github.com>:

> I can implement this, but that will be first week of October.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/transientskp/tkp/issues/456#issuecomment-143189993>.
>



-- 
Gijs Molenaar
http://pythonic.nl

**Comment by bartscheers** [comment]
Great!

**Comment by gijzelaerr** [comment]
@bartscheers i'm not sure about point 4. wouldn't that mean imply that the forcedfits count is always reset for every source association, not just for the blind fits ones?

**Comment by bartscheers** [comment]
Good to recheck...
No, I don't think so, because the nuldetections and monitoringlist_entries follow a different path than the blind detections. 
The list returned from get_nulldetections() should not include the sources with too many forced fits. These excluded sources should not end up in one of the association paths.

As a consequence: Do agree that there will be extractedsources (only of type forced fits) that will not end up in any association with an running catalogue source? Up till now all sources in were somehow associated.

**Comment by gijzelaerr** [comment]
sorry I don't get your question. Do you ask me for my opinion? can you rephrase your question?

**Comment by bartscheers** [comment]
The fact that we skip a forced-fit extractedsource from the association procedure means that there will be sources in the extractedsource table that are not associated with any source. Up till now, we made sure that every source in extractedsource had a counterpart in runningcatalog, but that won't be the case anymore.
That's the consequence of this issue. Maybe the question was rhetorical, let me rephrase the statement. We should be aware that there are unassociated sources in the extractedsource table.

**Comment by gijzelaerr** [comment]
how come? if it is a blind fits it will be associated with the running catalog right? And there will be only forced fits added to extractedsource when the forcedfits_count in the runningcatalog table is below the expiration thresshold.

**Comment by bartscheers** [comment]
Ok, you're right, sorry for the confusion. My head was too much in the association stuff, they won't be appended at all, forget my last comment.

**Comment by gijzelaerr** [comment]
this PR has been merged
